airbyte-cdk/python/airbyte_cdk/utils/datetime_format_inferrer.py

```diff
@@ -0,0 +1,80 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+from typing import Any, Dict, Union
+
+from airbyte_cdk.models import AirbyteRecordMessage
+from airbyte_cdk.sources.declarative.datetime.datetime_parser import DatetimeParser
+
+
+class DatetimeFormatInferrer:
+    """
+    This class is used to detect toplevel fields in records that might be datetime values, along with the used format.
+    """
+
+    def __init__(self):
+        self._parser = DatetimeParser()
+        self._datetime_candidates: Union[None, Dict[str, str]] = None
+        self._formats = [
+            "%Y-%m-%d",
+            "%Y-%m-%d %H:%M:%S",
+            "%Y-%m-%d %H:%M:%S.%f+00:00",
+            "%Y-%m-%dT%H:%M:%S.%f%z",
+            "%s",
+            "%d/%m/%Y %H:%M",
+            "%Y-%m",
+            "%d-%m-%Y",
+            "%Y-%m-%dT%H:%M:%SZ",
+        ]
+        self._timestamp_heuristic_range = range(1000000000, 2000000000)
+
+    def _can_be_datetime(self, value: Any) -> bool:
+        """Checks if the value can be a datetime. This is separate from the format check for performance reasons"""
+        if isinstance(value, str) and (not value.isdecimal() or int(value) in self._timestamp_heuristic_range):
+            return True
+        if isinstance(value, int) and value in self._timestamp_heuristic_range:
+            return True
+        return False
+
+    def _matches_format(self, value: Any, format: str) -> bool:
+        """Checks if the value matches the format"""
+        try:
+            self._parser.parse(value, format)
+            return True
+        except Exception:
```

Comment:
Should we use a more specific error to catch which is usually a better practice? A cursory test if we use bad inputs like a string ends up with a `ValueError`.